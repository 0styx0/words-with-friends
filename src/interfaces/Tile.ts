export default interface Tile {
    points: number,
    letter: string
}