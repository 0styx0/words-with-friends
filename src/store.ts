import reducers from './reducers';
import { createStore, compose } from 'redux';
import Player from './classes/Player';
import Tilebag from './classes/Tilebag';
import Board from './classes/Board';

const enhancers = compose('devToolsExtension' in window ?
    (window as {} as { devToolsExtension: Function }).devToolsExtension!() :
    (f: {}) => f);

export const defaultState = {
    turn: 0,
    Players: [new Player(true, 0), new Player(false, 1)], // turns switched around on INIT_PLAYERS
    board: new Board(),
    Tilebag: new Tilebag()
};

const store = createStore(reducers, defaultState, enhancers);

export default store;
