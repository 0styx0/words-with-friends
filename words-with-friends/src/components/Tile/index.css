.tile {
    background-color: #F3D32C;
    height: 100%;
    width: 100%;
    padding: 5px;
    border-radius: 12px;
    -moz-box-shadow: 1px 0px 2px 1px #D99536;
    -webkit-box-shadow: 1px 0px 2px 1px #D99536;;
    box-shadow: 1px 0px 2px 1px #D99536;;

}

.tile .points {
    float: right;
    font-size: 1.2vw;
    font-size: calc(7pt + 1.2vw);
}

.tile .letter {
    font-size: 2rem;
    position: relative;
    top: 30%;
}
