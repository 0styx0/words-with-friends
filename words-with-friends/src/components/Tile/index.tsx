import * as React from 'react';
import './index.css';
import Game from '../Game/Game';
import TileType from '../../interfaces/Tile';

interface Props {
    tile: TileType;
    coordinates?: string;
    canDrag?: boolean;
}

function onWildcardChange(event: any, coordinates?: string) {

    if (!coordinates) {
        return;
    }
    const tileInfo = Game.board.get(coordinates)!;

    if (tileInfo.tile!.points > 0 || !tileInfo.recent) {

        if (tileInfo) {
            event.target.value = tileInfo.tile!.letter;
            event.target.disabled = true;
        }
        return;
    }

    tileInfo.tile = { letter: event.target.value, points: 0, playerIndex: tileInfo.tile.playerIndex };

    Game.board.set(coordinates, tileInfo);
}

export default function Tile(props: Props) {

    const tileInfo = (props.coordinates && Game.board.get(props.coordinates)!) || {} as any;

    return (
        <div
            className="tile"
        >
            <span className="points">{props.tile.points}</span>
            {props.tile.letter === '' ?
                <select
                    disabled={!(!!props.canDrag && !!tileInfo && tileInfo.canDrag)}
                    onChange={(event) => onWildcardChange(event, props.coordinates)}
                >
                    {new Array(26)
                        .fill('', 0, 26)
                        .map((elt, i) =>
                        <option key={i}>{String.fromCharCode(i + 65)}</option>)}
                </select>
                :
                <span
                    className="letter"
                >
                    {props.tile.letter}
                </span>
            }
        </div>
    );
}
