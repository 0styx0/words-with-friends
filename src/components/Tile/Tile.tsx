import * as React from 'react';
import { DragEvent } from 'react';
import TileType from '../../interfaces/Tile';
import Tile from './';
import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import actionCreators from '../../actions';
import { defaultState } from '../../store';


function mapStateToProps(state: typeof defaultState, props: Props) {

    return {
        coordinates: props.coordinates,
        tile: props.tile,
        Players: state.Players,
        board: state.board,
        turn: state.turn
    };
}

function mapDispatchToProps(dispatch: Dispatch<typeof defaultState>) {
    return bindActionCreators(actionCreators, dispatch);
}


type Props = typeof actionCreators & typeof defaultState & {
    tile: TileType;
    removeTile: Function;
    coordinates: number[];
};


class TileContainer extends React.Component<Readonly<Props>, {}> {

    constructor(props: Readonly<Props>) {
        super(props);

        this.onDragStart = this.onDragStart.bind(this);
        this.onDragEnd = this.onDragEnd.bind(this);
    }

    /**
     * Sets event drop data to stringified `this.props.tile`
     */
    onDragStart(e: DragEvent<HTMLDivElement>) {
        e.dataTransfer.dropEffect = 'move';
        e.dataTransfer.setData('tile', JSON.stringify(this.props.tile));
    }

    /**
     * If tile was dropped somewhere else (checked using `dropEffect`), removes tile from this TileContainer
     *  (so the tile doesn't get duplicated, just moved)
     */
    onDragEnd(e: DragEvent<HTMLDivElement>) {

        const tileWasMoved = e.dataTransfer.dropEffect !== 'none';

        if (tileWasMoved) {
            this.props.removeTile();
        }
    }

    /**
     * @return true if tile is already on board and was placed on current turn,
     *  or tile is in current player's hand. Else false
     */
    canDrag() {

        const data = this.props.board.get(this.props.coordinates);

        const tileIsAlreadyOnBoard = data && !!data!.tile;

        const tileInCurrentPlayersHand =
            !tileIsAlreadyOnBoard && this.props.Players[this.props.tile.playerIndex!].turn;
        const draggableTileOnBoard = tileIsAlreadyOnBoard && data!.canDrag;

        return !!(tileInCurrentPlayersHand || draggableTileOnBoard);
    }

    render() {

        const canDrag = this.canDrag();

        return (
            <div
                className="tileContainer"
                onDragStart={this.onDragStart}
                onDragEnd={this.onDragEnd}
                draggable={canDrag}
            >
                <Tile
                    tile={this.props.tile}
                    coordinates={this.props.coordinates}
                    canDrag={canDrag}
                />
            </div>
        );
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(TileContainer as any);
