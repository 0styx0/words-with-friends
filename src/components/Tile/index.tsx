import * as React from 'react';
import './index.css';
import TileType from '../../interfaces/Tile';
import store, { getState } from '../../store';
import { ChangeEvent } from 'react';

interface Props {
    tile: TileType;
    coordinates?: number[];
    canDrag?: boolean;
}

function onWildcardChange(event: ChangeEvent<HTMLSelectElement>, coordinates?: number[]) {

    if (!coordinates) {
        return;
    }

    const board = getState().board;
    const tileInfo = board.get(coordinates)!;

    if (tileInfo.tile!.points > 0 || !tileInfo.recent(getState().turn)) {

        if (tileInfo) {
            event.target.value = tileInfo.tile!.letter;
            event.target.disabled = true;
        }
        return;
    }

    store.dispatch({
        type: 'PLACE_TILE_ON_BOARD',
        tile: {
            letter: event.target.value,
            points: 0,
            playerIndex: tileInfo.tile!.playerIndex
        },
        coordinates
    });
}

export default function Tile(props: Props) {

    return (
        <div
            className="tile"
        >
            <span className="points">{props.tile.points}</span>
            {props.tile.points === 0 ?
                <select
                    disabled={!props.canDrag}
                    onChange={(event) => onWildcardChange(event, props.coordinates)}
                >
                    {new Array(26)
                        .fill('', 0, 26)
                        .map((elt, i) =>
                        <option key={i}>{String.fromCharCode(i + 65)}</option>)}
                </select>
                :
                <span
                    className="letter"
                >
                    {props.tile.letter}
                </span>
            }
        </div>
    );
}
