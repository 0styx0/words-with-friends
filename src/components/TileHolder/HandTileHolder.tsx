import * as React from 'react';
import TileHolder from './';
import TileType from '../../interfaces/Tile';
import AbstractTileHolder from './AbstractTileHolder';

import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import actionCreators from '../../actions';
import { defaultState } from '../../store';
import Player from '../../classes/Player';
// import Computer from '../../classes/Computer';
import Board from '../../classes/Board';


function mapStateToProps(state: typeof defaultState, props: Props) {
    return {
        tile: props.tile,
        Players: state.Players,
        board: state.board,
        turn: state.turn
    };
}

function mapDispatchToProps(dispatch: Dispatch<typeof defaultState>) {
    return bindActionCreators(actionCreators, dispatch);
}


type Props = typeof actionCreators & typeof defaultState & {
    tile?: TileType;
    Players: Player;
    board: Board;
    turn: number;
};


export class HandTileHolderContainer extends AbstractTileHolder<Props> {

    /*
    placeWord(computer: Computer) {
        
        const highestWord = computer.getHighestWord(
             computer.getPossibleWords(this.props.board, this.props.turn), this.props.board, this.props.turn
         );

        if (highestWord.horizontal) {

            for (
                let i = highestWord.startCoordinate[0];
                i < highestWord.startCoordinate[0] + highestWord.length;
                i++
            ) {
                this.putTile(computer.tiles.find(tile => tile.letter === highestWord.word[i])!);
            }
        }
    }
    */

    /**
     * Puts tile down (@see tileTarget.drop)
     */
    putTile(tile: TileType) {

        this.props.putTileInHand(this.props.Players, tile!);
    }

    /**
     * Removes tile from TileHolder (@see Tile.tsx#source.endDrag)
     */
    removeTile() {
        this.props.removeTileFromHand(this.props.Players, this.props.tile!);
    }

    render() {

        return (
            <div
              className="tileHolder"
              onDrop={this.onDrop}
              onDragOver={this.onDragOver}
            >
                <TileHolder
                    tile={this.props.tile}
                    removeTile={this.removeTile}
                />
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(HandTileHolderContainer as any);
