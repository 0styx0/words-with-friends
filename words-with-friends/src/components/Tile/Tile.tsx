import * as React from 'react';
import Tile from './';
import { DragSource } from 'react-dnd';
import Game from '../../classes/Game';

interface Props {
    letter: string;
    points: number;
    connectDragSource?: Function;
    isDragging?: Function;
    removeTile: Function;
    canDrag: boolean;
    coordinates: string;
}

interface State {
    stop: boolean;
}

function collect(connect: {dragSource: Function}, monitor: {isDragging: Function}) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  };
}

const source = {

  canDrag(props: Props) {

        const data = Game.board.get(props.coordinates);

        if (!data) {
            return true;
        }

        return props.canDrag && (data.turnTileWasPlaced === Game.turn || data.turnTileWasPlaced === 0);
  },

  beginDrag(props: Props) {
    return props;
  },
  endDrag(props: Props, monitor: any, senderComponent: any) {

      if (monitor.didDrop()) {
          senderComponent.props.removeTile();
      }
  }
};

@DragSource('tile', source, collect)
export default class TileContainer extends React.Component<Props, State> {

    constructor() {
        super();
    }

    render() {

        const { connectDragSource, isDragging } = this.props as {connectDragSource: Function, isDragging: Function };

        return connectDragSource(
            <div
              style={{
                opacity: isDragging ? 0.99 : 1,
                fontSize: 25,
                fontWeight: 'bold',
                cursor: 'move'
              }}
            >
                <Tile {...this.props} />
            </div>
        );
    }

}