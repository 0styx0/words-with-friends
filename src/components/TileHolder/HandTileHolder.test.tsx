import * as React from 'react';
import * as renderer from 'react-test-renderer';
import store, { defaultState, getState } from '../../store';
import { Provider } from 'react-redux';
import HandTileHolder, { HandTileHolderContainer } from './HandTileHolder';
import mockMath from '../../test/mocks/Math';
import { mount } from 'enzyme';
import Tile from '../../interfaces/Tile';
mockMath();

describe('<HandTileHolder />', () => {

    describe('renders', () => {

        function snap(playerIndex: number) {

            const state = getState();

            const tree = renderer.create(
                <Provider store={store}>
                    <HandTileHolder tile={state.Tilebag.getRandomTile(playerIndex)} {...[] as any} />
                </Provider>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it(`correctly for current player's tile`, () => snap(0));

        it(`correctly for other player's tile`, () => snap(1));
    });

    function setup(tile: Tile = getState().Tilebag.getRandomTile(0)) {

        const wrapper = mount(

            <Provider store={store}>
                <HandTileHolder tile={tile} {...[] as any} />
            </Provider>
        );

        return {
            wrapper,
            component: wrapper.find(HandTileHolderContainer).instance() as HandTileHolderContainer
        };
    }

    describe('#putTile', () => {

        it('adds tile to hand', () => {

            const initialTile = { points: -1, letter: 'invalid letter', playerIndex: 0 };
            const { component } = setup(initialTile);

            const currentPlayerTiles = getState().Players.find(player => player.turn)!.tiles;

            expect(currentPlayerTiles).toHaveLength(0);
            expect(component.props.tile).toEqual(initialTile);

            const newTile = getState().Tilebag.getRandomTile(0);

            component.putTile(newTile);

            expect(currentPlayerTiles).toEqual([newTile]);
        });
    });

    describe('#removeTile', () => {

        it('removes tile as a prop', () => {

            const currentPlayerTiles = getState().Players.find(player => player.turn)!.tiles;
            expect(currentPlayerTiles).toHaveLength(1);

            const { component } = setup(currentPlayerTiles[0]);

            component.removeTile();

            const currentPlayerUpdatedTiles = getState().Players.find(player => player.turn)!.tiles;
            expect(currentPlayerUpdatedTiles).toHaveLength(0);
        });
    });
});