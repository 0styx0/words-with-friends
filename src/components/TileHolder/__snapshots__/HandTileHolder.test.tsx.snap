// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HandTileHolder /> renders correctly for current player's tile 1`] = `
<div
  className="tileHolder"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div
    className="tileContainer"
    draggable={true}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="tile"
    >
      <span
        className="points"
      >
        2
      </span>
      <span
        className="letter"
      >
        L
      </span>
    </div>
  </div>
</div>
`;

exports[`<HandTileHolder /> renders correctly for other player's tile 1`] = `
<div
  className="tileHolder"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div
    className="tileContainer"
    draggable={false}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="tile"
    >
      <span
        className="points"
      >
        5
      </span>
      <span
        className="letter"
      >
        K
      </span>
    </div>
  </div>
</div>
`;
